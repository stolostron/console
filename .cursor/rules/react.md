# React Rules for ACM Console

## Component Patterns

### Functional Components
- Use functional components with hooks exclusively
- Implement proper prop interfaces with TypeScript
- Use React.memo for expensive components
- Follow established naming conventions (PascalCase)

### Hook Usage
- Use custom hooks for business logic separation
- Implement proper dependency arrays in useEffect
- Use useCallback for event handlers passed as props
- Leverage useMemo for expensive computations

### State Management
- Use Recoil atoms through useSharedAtoms() pattern
- Access atoms with useRecoilValue() for reading
- Use useRecoilValueGetter() for async operations
- Follow established atom naming conventions

Example (based on actual codebase patterns):
```typescript
/* Copyright Contributors to the Open Cluster Management project */
import { useMemo } from 'react'
import { useRecoilValue } from 'recoil'
import { useTranslation } from '../lib/acm-i18next'
import { useSharedAtoms } from '../shared-recoil'
import { AcmTable } from '../ui-components'

export function ClusterList() {
  const { t } = useTranslation()
  const { managedClustersState } = useSharedAtoms()
  const clusters = useRecoilValue(managedClustersState)
  
  const filteredClusters = useMemo(
    () => clusters.filter(cluster => cluster.status?.conditions),
    [clusters]
  )
  
  return (
    <AcmTable
      items={filteredClusters}
      columns={clusterColumns}
      keyFn={cluster => cluster.metadata.uid}
    />
  )
}
```

### Conditional Rendering
- Use logical AND (`&&`) for simple conditional rendering (ACM standard pattern):
  ```typescript
  {condition && <Component />}
  {props.additionalLabels && (<span><AcmLabels /></span>)}
  {expandable && itemCount > columnCount && (<AcmButton />)}
  ```
- Use ternary operators for either/or scenarios:
  ```typescript
  {props.isCompact ? <Popover /> : renderLabelGroup()}
  {showAll ? t('Show less') : t('Show all')}
  ```
- Early returns for major conditional logic:
  ```typescript
  if (props.hidden) return <></>
  if (!data) return <LoadingState />
  ```

### Performance
- Use `React.memo` for components that receive the same props frequently
- Use `useMemo` for expensive calculations
- Use `useCallback` for functions passed to child components
- Implement lazy loading for heavy components
- Avoid inline object/array creation in render
- Avoid using index as the key prop value
- Avoid using inline functions in JSX or TSX blocks
- Use proper key props for list items

### Error Boundaries
- Implement error boundaries for route-level components
- Use proper error state management
- Provide user-friendly error messages
- Log errors for debugging purposes

### Accessibility
- Use semantic HTML elements
- Implement proper ARIA attributes
- Ensure keyboard navigation works
- Test with screen readers

### Component Organization
- Keep components small and focused on a single responsibility
- Use PascalCase for component names and component files
- Use descriptive component names that indicate the component's purpose
- Avoid abbreviations in component names
- Separate presentation from business logic
- Use proper file structure (components/, ui-components/, routes/)
- Export components properly with named exports
- Avoid props drilling where props are only passed down to child components
- Do not mutate props or state - create a copy array or object if possible
