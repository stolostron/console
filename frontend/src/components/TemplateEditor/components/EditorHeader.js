/* Copyright Contributors to the Open Cluster Management project */
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { Alert, AlertActionCloseButton, ClipboardCopy, Checkbox } from '@patternfly/react-core'
import '../css/editor-header.css'

class EditorHeader extends React.Component {
  static propTypes = {
    children: PropTypes.node,
    handleEditorCommand: PropTypes.func,
    handleShowSecretChange: PropTypes.func,
    handleTabChange: PropTypes.func,
    i18n: PropTypes.func,
    otherYAMLTabs: PropTypes.array,
    readOnly: PropTypes.bool,
    showSecrets: PropTypes.bool,
    title: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
  }

  constructor(props) {
    super(props)
  }

  render() {
    const { children, readOnly, otherYAMLTabs = [], title, handleEditorCommand, i18n } = this.props
    const editorToolbarTitle = i18n('editor.toolbar')
    if (readOnly) {
      return (
        <div>
          <Alert
            isInline
            title={i18n('editor.bar.readonly')}
            variant={'info'}
            style={{ background: '#E7F1FA', padding: '15px 20px' }}
            actionClose={<AlertActionCloseButton onClose={() => handleEditorCommand('close')} />}
          />
          <div className="readonly-editor-bar">
            <div>{title}</div>
            <ClipboardCopy variant="inline-compact" isBlock onCopy={() => handleEditorCommand('copyAll')} />
          </div>
        </div>
      )
    } else {
      const hasTabs = otherYAMLTabs.length > 0
      const classnames = classNames({
        'creation-view-yaml-header': true,
        hasTabs: hasTabs,
      })
      return (
        <div className={classnames}>
          <div
            className="creation-view-yaml-header-toolbar"
            role="region"
            aria-label={editorToolbarTitle}
            id={editorToolbarTitle}
          >
            {children}
          </div>
          <div className="creation-view-yaml-header-tabs">
            {this.renderEditorTabs(otherYAMLTabs)}
            {this.renderShowSecrets()}
            <ClipboardCopy variant="inline-compact" isBlock onCopy={() => handleEditorCommand('copyAll')} />
          </div>
        </div>
      )
    }
  }

  setTabsRef = (ref) => {
    this.tabsRef = ref
  }

  renderEditorTabs = (otherYAMLTabs) => {
    const { type = 'unknown', handleTabChange } = this.props

    const onClick = (e, tab) => {
      e.preventDefault()
      Array.from(this.tabsRef.children).forEach((child, inx) =>
        child.classList.toggle('tf--tabs__nav-item--selected', inx === tab)
      )
      handleTabChange(tab)
    }
    return (
      <nav aria-label="Select template" className="tf--tabs" role="navigation">
        <ul role="tablist" className="tf--tabs__nav" ref={this.setTabsRef}>
          <li
            id="main"
            role="presentation"
            tabIndex="-1"
            className="tf--tabs__nav-item tf--tabs__nav-item--selected"
            onClick={(e) => onClick(e, 0)}
          >
            <a className="tf--tabs__nav-link" href="#" role="tab" tabIndex="0" aria-selected="true">
              {type}
            </a>
          </li>
          {otherYAMLTabs.map(({ id }, inx) => {
            return (
              <li
                id={id}
                key={id}
                role="presentation"
                tabIndex="-1"
                className="tf--tabs__nav-item"
                onClick={(e) => onClick(e, inx + 1)}
              >
                <a className="tf--tabs__nav-link" href="#" role="tab" tabIndex="0" aria-selected="false">
                  {id}
                </a>
              </li>
            )
          })}
        </ul>
      </nav>
    )
  }

  renderShowSecrets = () => {
    const { showSecrets, handleShowSecretChange, i18n } = this.props
    return (
      <div className="creation-view-yaml-header-secrets">
        <Checkbox
          aria-label="show-secrets"
          id="show-secrets"
          isChecked={showSecrets}
          onChange={handleShowSecretChange}
        />
        <div>{i18n('editor.show.secrets')}</div>
      </div>
    )
  }
}

export default EditorHeader
