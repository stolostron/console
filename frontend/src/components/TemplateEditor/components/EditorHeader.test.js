/* Copyright Contributors to the Open Cluster Management project */
'use strict'

import React from 'react'
import EditorHeader from './EditorHeader'
import { render } from '@testing-library/react'

const type = 'application'
const otherYAMLTabs = []
const handleTabChange = jest.fn

describe('EditorHeader component', () => {
    it('renders as expected', () => {
        const fn = jest.fn()
        const Component = () => {
            return (
                <EditorHeader
                    otherYAMLTabs={otherYAMLTabs}
                    handleTabChange={handleTabChange}
                    type={type}
                    i18n={fn}
                    title=""
                />
            )
        }
        const { asFragment } = render(<Component />)
        expect(asFragment()).toMatchSnapshot()
    })
})
