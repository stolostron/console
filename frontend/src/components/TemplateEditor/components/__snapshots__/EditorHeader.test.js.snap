// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorHeader component renders as expected 1`] = `
<DocumentFragment>
  <div
    class="creation-view-yaml-header"
  >
    <div
      class="creation-view-yaml-header-toolbar"
      role="region"
    />
    <div
      class="creation-view-yaml-header-tabs"
    >
      <nav
        aria-label="Select template"
        class="tf--tabs"
        role="navigation"
      >
        <ul
          class="tf--tabs__nav"
          role="tablist"
        >
          <li
            class="tf--tabs__nav-item tf--tabs__nav-item--selected"
            id="main"
            role="presentation"
            tabindex="-1"
          >
            <a
              aria-selected="true"
              class="tf--tabs__nav-link"
              href="#"
              role="tab"
              tabindex="0"
            >
              application
            </a>
          </li>
        </ul>
      </nav>
      <div
        class="creation-view-yaml-header-secrets"
      >
        <div
          class="pf-c-check pf-m-standalone"
        >
          <input
            aria-invalid="false"
            aria-label="show-secrets"
            class="pf-c-check__input"
            data-ouia-component-id="OUIA-Generated-Checkbox-1"
            data-ouia-component-type="PF4/Checkbox"
            data-ouia-safe="true"
            id="show-secrets"
            type="checkbox"
          />
        </div>
        <div />
      </div>
      <div
        class="pf-c-clipboard-copy pf-m-inline pf-m-block"
      >
        <span
          class="pf-c-clipboard-copy__text"
          id="text-input-0"
        />
        <span
          class="pf-c-clipboard-copy__actions"
        >
          <span
            class="pf-c-clipboard-copy__actions-item"
          >
            <button
              aria-disabled="false"
              aria-label="Copy to clipboard"
              aria-labelledby="copy-button-0 text-input-0"
              class="pf-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              id="copy-button-0"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 448 512"
                width="1em"
              >
                <path
                  d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
                />
              </svg>
            </button>
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
